### PR TITLE
Add support for editing user notes

### PR DESCRIPTION
**Pull request by ghost**
This adds functionality for the endpoint to edit a user note. This is a note
that is user-specific (a user is the only one that can see a note they make
about another user).

This is functionality that can not be used by bots, according to Jake. source:
https://github.com/hydrabolt/discord.js/issues/327#issuecomment-220796346

This is not something listed in the official API docs, but it is a feature
utilized by the desktop client.

edit:

as setting the note to `""` is equivilant to removing the note (no, there's not a DELETE method for this endpoint), I was wondering if changing `note: &str` to `note: Option<&str>` was preferred, to signify that no note is equivilant to deleting it. Or perhaps it just needs to be in the function's docs that an empty note is the equivilant of deletion.

### CODE REVIEW

**Comment by SpaceManiac** [comment]
Take a `UserId` rather than a `&UserId`. Taking `&str` is fine if the docs state that an empty string deletes the note.

**Comment by ghost** [comment]
Updated to `UserId`.

The funny thing about notes is that they aren't anywhere in the docs yet, albeit being around for at least 5 weeks. The docs repo only contains [1 reference to notes](https://github.com/hammerandchisel/discord-api-docs/blame/6f49afcbcf36de340f991b92fbffe58ce9604591/docs/topics/RESPONSE_CODES.md#L64), and that's for a JSON error code.

Perhaps this needs to be listed in the function docs that an empty string is the equivilant.

In addition, I've made a couple changes that's a first for me, so I'm unsure if I got it right but I think I did - everything works fine for me in my tests.

Since I added the function to create/update a note, I figured that the client must be getting the notes from somewhere. I looked around and it looks like it's given on the READY payload. I've updated the ready event decoder to parse the notes, and I've also created an event for a new event from the ws, `USER_NOTE_UPDATE`.

File: src/state.rs [hunk 1]
```diff
@@ -306,6 +317,9 @@ impl State {
 		}
 		None
 	}
+
+	/// Get the map of notes that have been made by this user.
+	pub fn notes(&self) -> &Option<BTreeMap<UserId, String>> { &self.notes }
```
Comment (SpaceManiac) [inline src/state.rs]: I think `Option<&...>` would be preferable (by returning `self.notes.as_ref()`)

File: src/model.rs [hunk 2]
```diff
@@ -1588,6 +1597,37 @@ fn decode_discriminator(value: Value) -> Result<String> {
...
+			Some(notes)
+		},
+		None => None,
+	};
+
+	if let Some(notes) = notes {
+		Ok(notes)
+	} else {
+		Err(Error::Decode("Error decoding notes", value))
+	}
```
Comment (SpaceManiac) [inline src/model.rs]: I think `notes.ok_or(Error::Decode(...))` would be clearer than this if-let.

File: src/model.rs [hunk 3]
```diff
@@ -1588,6 +1597,37 @@ fn decode_discriminator(value: Value) -> Result<String> {
 	}
 }
 
+fn decode_notes(value: Value) -> Result<BTreeMap<UserId, String>> {
+	let notes = match value.as_object() {
+		Some(object) => {
+			let mut notes: BTreeMap<UserId, String> = BTreeMap::new();
+
+			for (key, value) in object.iter() {
```
Comment (SpaceManiac) [inline src/model.rs]: If you use `.into_iter()` here, you can probably avoid the String copy at `String::from` later in this method. Use `into_string` if needed.

File: src/model.rs [hunk 4]
```diff
@@ -1313,6 +1316,7 @@ impl Event {
 				user_settings: try!(opt(&mut value, "user_settings", UserSettings::decode)),
 				user_server_settings: try!(opt(&mut value, "user_guild_settings", |v| decode_array(v, UserServerSettings::decode))),
 				tutorial: try!(opt(&mut value, "tutorial", Tutorial::decode)),
+				notes: try!(opt(&mut value, "notes", |v| decode_notes(v))),
```
Comment (SpaceManiac) [inline src/model.rs]: Writing `|v| decode_notes(v)` is redundant; you should be able to write `decode_notes` and get the same effect.

File: src/state.rs [hunk 5]
```diff
@@ -69,6 +72,14 @@ impl State {
 		match *event {
 			Event::Ready(ref ready) => *self = State::new(ready.clone()),
 			Event::UserUpdate(ref user) => self.user = user.clone(),
+			Event::UserNoteUpdate(ref user_id, ref note) => {
+				match self.notes.as_mut() {
+					Some(notes) => {
+						notes.insert(UserId(user_id.0), note.clone());
+					},
+					None => {},
```
Comment (SpaceManiac) [inline src/state.rs]: An if-let rather than a `match` would reduce nesting here, but it's a pretty minor stylistic concern.

**Comment by ghost** [comment]
So I addressed 4 of those, but I'm unsure about the comment regarding using `into_iter()`.

AFAICT, the `String::from` is necessary, because a serde Value string is a `&str`. Passing the Value to `into_string` would not work, because `into_string` only allows moving the value into it, whereas this would be borrowed. What's an implementation that would work without the `String::from`?

I can imagine getting the keys of the map, and then `get(&key)`ing them to get the underlying values, and then `into_string`ing them, but I'm not sure how expensive this is in comparison.

**Comment by SpaceManiac** [comment]
Ah, I see what you mean about the trickiness of avoiding the clone. The possibility is definitely there though, I'll take a stab at it myself.

**Comment by abalabahaha** [comment]
Docs don't reference notes because docs are intended for OAuth bot dev reference

**Comment by SpaceManiac** [comment]
Go ahead and squash and rebase, I'll commit the no-clone decoding separately.

**Comment by ghost** [comment]
Updated.

**Pull request merged**